File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low = 0 , high = nums.size()-1;
        while(low<=high){
            int  mid = low + (high-low)/2;
            if(nums[mid]==target){
                return mid;
            }
            else if(nums[mid]>= nums[low]){
                if(target>= nums[low] && target<=nums[mid]){
                    high = mid-1;
                }
                else low= mid+1;
            }
            else {
                if(target>=nums[mid] && target<=nums[high]){
                    low = mid+1;
                }
                else{
                    high = mid-1;
                }
            }
            
            
           
        }
        return -1;
    }
};

// int search(vector<int>& nums, int target) {
//     //there exists atleast one strictly increasing subarray due to pivot element
//     //using modified binary search 
//         int n=nums.size();
//         int s=0;
//         int e=n-1;
        
//         while(s<=e){
//             int mid=s+(e-s)/2;
            
//             if(nums[mid] == target) return mid;
//   //else if left subarray is str increasing then search only in the LEFT SUBARRAY          
//             else if(nums[mid]>=nums[s]){
//                 if(target>=nums[s] && target <=nums[mid]){
                    
//                     e =mid-1;//update the right pointer
                    
//                 }
//                 else
//                     s=mid+1;//update the left pointer
//             }
//  //else search in the RIGHT SUBARRAY
//             else{
//                 if(target>=nums[mid] && target<=nums[e]){
                    
//                     s=mid+1;
                    
//                 }
//                 else
//                     e=mid-1;
//             }
            
//         }
        
//         return -1;//if it is not PRESENT
//     }
// };

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int idx = minIndex(nums); 
        System.out.println(idx);
        int res1=-1, res2=-1;
        if(idx>=1) res1 = binarySearch(nums,0,idx-1,target);
        res2 = binarySearch(nums,idx,n-1,target);
        return res1==-1 ? res2 : res1;
        
    }
    
    public int binarySearch(int[] nums , int low , int high, int target){
        while(low<=high){
            int m= (low+high)/2;
            if(nums[m]==target)return m;
            else if(nums[m]>target) high= m-1;
            else low= m+1;
        }
        return -1;
    }
    
    public int minIndex(int[] nums){
        int low= 0 , high = nums.length-1;
//          two base cases
        if(low==high) return low;
        if(nums[low]< nums[high])return low;
        
        int ans = Integer.MAX_VALUE;
        int res = -1;
        while(low<=high){
            int m = (low+high)/2;
            if(nums[m]>=nums[0]){
                low = m+1;
            }
            else{
                if(nums[m]<ans){
                    ans=nums[m];
                    res=m;
                }
                
                high = m-1;
                
            }
        }
        return res;
    }
}